flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/TableSourceTable.scala

```diff
@@ -50,13 +53,29 @@ class TableSourceTable[T](
     this(tableSource, isStreamingMode, statistic, None, catalogTable)
   }
 
+  Preconditions.checkNotNull(tableSource)
+  Preconditions.checkNotNull(statistic)
+  Preconditions.checkNotNull(catalogTable)
+
+  val watermarkSpec: Option[WatermarkSpec] = catalogTable
+    .getSchema
+    .getWatermarkSpecs.asScala.headOption
+
+  if (TableSourceValidation.hasRowtimeAttribute(tableSource) && watermarkSpec.isDefined) {
```

Comment:
This will prevent `kafka` table from registering watermark through DDL.

Reply:
Nope. Only when users specify rowtime attributes explicitly via `Rowtime` descriptor API, the KafkaTableSource will return non-empty `RowtimeAttributeDescriptor`.  

It's fine if a `TableSource` implements `DefinedRowtimeAttributes` but returns empty `RowtimeAttributeDescriptor` and defines watermark in DDL at the same time. 

I will improve the exception message a bit more.